### PR TITLE
FM2-80: Improve Search for Location Resource

### PR DESCRIPTION
**Pull request by varung-31**
## Description of what I changed

1. Added a single search method for Location resource containing name, city, state, postal code, country, tags.
2. Implemented the "part of" functionality as well to the above.
3. Added sorting by name, city, country, postal code, state.

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-80

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**.

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by varung-31** [comment]
> For the DAO implementation, I believe most of the functionality is already implemented in the Base DAO, such as `handleNames`, `handleAddress`. This would be similar to how it is implemented in Person/Patient DAO.

Yeah but there address is implemented as Optional.of(eq(...)) and I want it to be Optional.of(ilike(...))
I'll change the handleNames though.

**Comment by varung-31** [comment]
> > For the DAO implementation, I believe most of the functionality is already implemented in the Base DAO, such as `handleNames`, `handleAddress`. This would be similar to how it is implemented in Person/Patient DAO.
> 
> Yeah but there address is implemented as Optional.of(eq(...)) and I want it to be Optional.of(ilike(...))
> I'll change the handleNames though.

@ibacher @CaptainDaVinci  Isn't it better to implement BaseDaoImpl using ilike instead of eq ??

**Comment by CaptainDaVinci** [comment]
@varung-31 This was discussed on talk as well ([link](https://talk.openmrs.org/t/intended-behavior-of-search-in-fhir-person-patient-resource/27025/2)).

**Comment by CaptainDaVinci** [comment]
@varung-31 actually, `handleNames` is used for searching human name and not location name and that's not what we want here, I think.

**Comment by varung-31** [comment]
> @varung-31 actually, `handleNames` is used for searching human name and not location name and that's not what we want here, I think.

@CaptainDaVinci I figured the same thing a while ago. Thanks though.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 1]
```diff
@@ -30,49 +39,99 @@
...
+			Set<Object> possibleParentLocationIdList = new HashSet<>(criteriaForParent.setProjection(projList).list());
+
+			criteria.createCriteria("parentLocation").add(in("locationId", possibleParentLocationIdList));
+		}
+	}
+
 	@Override
-	public Collection<Location> findLocationsByState(String state) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("stateProvince", state)).list();
+	protected String paramToProp(@NotNull String paramName) {
```
[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: I have already done that for "name". In this PR, I only sorted by name.

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 2]
```diff
@@ -36,166 +43,258 @@
...
+				null, null, null, sort);
+
+		assertThat(results, notNullValue());
+		assertThat(results, not(empty()));
+		assertThat(results.size(), greaterThan(1));
+
+		resultsList = new ArrayList<>(results);
+		// check if the sorting is indeed correct by descending order
+		for (int i = 1; i < resultsList.size(); i++) {
+			assertTrue(resultsList.get(i - 1).getName().compareTo(resultsList.get(i).getName()) >= 0);
```
[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Done

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 3]
```diff
@@ -36,166 +43,258 @@
...
+				null, null, null, sort);
+
+		assertThat(results, notNullValue());
+		assertThat(results, not(empty()));
+		assertThat(results.size(), greaterThan(1));
+
+		List<Location> resultsList = new ArrayList<>(results);
+		// check if the sorting is indeed correct by ascending order
+		for (int i = 1; i < resultsList.size(); i++) {
+			assertTrue(resultsList.get(i - 1).getName().compareTo(resultsList.get(i).getName()) <= 0);
```
[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Done

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 4]
```diff
@@ -36,166 +43,258 @@
...
-	
+
 	private static final String UNKNOWN_POSTAL_CODE = "4015-3100";
-	
+
 	private static final String LOCATION_STATE = "province";
-	
+
 	private static final String UNKNOWN_LOCATION_STATE = "province state";
-
```
[the remark this answers is not in the document]
Comment (varung-31) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Done

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java [hunk 5]
```diff
@@ -693,19 +697,27 @@ protected String paramToProp(@NotNull String param) {
...
+		
+		return Optional.of(ilike(propertyName, param.getValue(), MatchMode.START));
+	}
+	
+	protected Optional<Criterion> propertyLike(@NotNull String propertyName, TokenParam tag) {
+		if (tag == null) {
+			return Optional.empty();
+		}
+		
+		return propertyLike(propertyName, new StringParam(tag.getValue()));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: I don't think we should actually be adding this. For `TokenParam` unlike `StringParam`, the match should always be equality, i.e., so instead of `propertyLike("t.uuid", tag)`, we should really just be doing `eq("t.uuid", tag)`.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 6]
```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
...
+	private void handleState(Criteria criteria, StringParam state) {
+		if (state != null) {
+			propertyLike("stateProvince", state).ifPresent(criteria::add);
+		}
+	}
+	
+	private void handleTag(Criteria criteria, TokenParam tag) {
+		if (tag != null) {
+			criteria.createAlias("tags", "t");
+			propertyLike("t.name", tag).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: See above.

```java
critieria.add(eq("t.name", tag.getValue()))
```

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 7]
```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
...
+		}
+	}
+	
+	// handle queries of the form /Location?partof=parent_name
+	private void handleParentLocation(Criteria criteria, StringParam parent) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: `parent` should be defined as a `referenceParam`. See the definition [here](https://www.hl7.org/fhir/location.html#search).

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 8]
```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
...
+		if (parent != null) {
+			ProjectionList projList = Projections.projectionList();
+			projList.add(Projections.property("locationId"));
+			
+			Criteria criteriaForParent = this.sessionFactory.getCurrentSession().createCriteria(Location.class);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: If you're going to do something like this, used `DetachedCriteria` instead of `Criteria`.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 9]
```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
...
+			propertyLike("name", parent).ifPresent(criteriaForParent::add);
+			
+			Set<Object> possibleParentLocationIdList = new HashSet<>(criteriaForParent.setProjection(projList).list());
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: Instead of this:

```java
criteriaForParent.setProjection(projectionList);

criteria.add(Subqueries.in("parent.locationId", criteriaForParent));
```

Your current implementation results in 2 db queries. This results in just one, which should be our goal.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 10]
```diff
@@ -30,7 +39,7 @@
 
 @Component
 @Setter(AccessLevel.PACKAGE)
-public class FhirLocationDaoImpl implements FhirLocationDao {
+public class FhirLocationDaoImpl extends BaseDaoImpl implements FhirLocationDao {
 	
 	@Inject
 	LocationService locationService;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: `private LocationService locationService`

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 11]
```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
...
+			default:
+				return null;
+		}
 	}
 	
 	@Override
-	public Collection<Location> findLocationsByTag(TokenParam tag) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).createAlias("tags", "t")
-		        .add(eq("t.name", tag.getValue())).list();
+	public Collection<Location> searchForLocations(StringParam name, StringParam city, StringParam country,
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: As this is a `public` method, it should be near the top of the class.

File: omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java [hunk 12]
```diff
@@ -56,37 +58,13 @@ public Location getLocationById(@IdParam @NotNull IdType id) {
...
-	@SuppressWarnings("unused")
-	public Bundle findLocationByState(@RequiredParam(name = Location.SP_ADDRESS_STATE) StringParam state) {
-		return FhirUtils.convertSearchResultsToBundle(fhirLocationService.findLocationsByState(state.getValue()));
-	}
-	
-	@Search
-	@SuppressWarnings("unused")
-	public Bundle findLocationsByTag(@RequiredParam(name = "_tag") TokenParam tag) {
-		return FhirUtils.convertSearchResultsToBundle(fhirLocationService.findLocationsByTag(tag));
+	public Bundle searchLocations(@OptionalParam(name = Location.SP_NAME) StringParam name,
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java]: These should be `StringOrListParam`s

File: omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java [hunk 13]
```diff
@@ -56,37 +58,13 @@ public Location getLocationById(@IdParam @NotNull IdType id) {
...
+	        @OptionalParam(name = Location.SP_ADDRESS_CITY) StringParam city,
+	        @OptionalParam(name = Location.SP_ADDRESS_COUNTRY) StringParam country,
+	        @OptionalParam(name = Location.SP_ADDRESS_POSTALCODE) StringParam postalCode,
+	        @OptionalParam(name = Location.SP_ADDRESS_STATE) StringParam state, @OptionalParam(name = "_tag") TokenParam tag,
+	        @OptionalParam(name = Location.SP_PARTOF) StringParam parent, @Sort SortSpec sort) {
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java]: This should be a `ReferenceParam` or, better, a `ReferenceOrListParam`. Note you can add searches for additional properties as we do for `Patient` elsewhere.

**Comment by VaishSiddharth** [comment]
@varung-31 I guess you have some conflicts with the master branch. You need to resolve them before this can be merged into the master branch. You can try using `git pull --rebase upstream master` to rebase to master

**Comment by varung-31** [comment]
> @varung-31 I guess you have some conflicts with the master branch. You need to resolve them before this can be merged into the master branch. You can try using `git pull --rebase upstream master` to rebase to master

I know @VaishSiddharth. I still need to publish a new revision. I'll pull the changes in that.

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 14]
```diff
@@ -101,91 +113,140 @@ public void getLocationByUuid_shouldReturnNullWithUnknownUuid() {
...
-	public void findLocationsByTag_shouldReturnLocationsContainingGivenTag() {
+	public void searchForLocations_shouldReturnLocationsContainingGivenTag() {
 		TokenParam locationTag = new TokenParam(FhirConstants.OPENMRS_FHIR_EXT_LOCATION_TAG, LOGIN_LOCATION_TAG_NAME);
-		Collection<Location> locations = fhirLocationDao.findLocationsByTag(locationTag);
+		Collection<Location> locations = fhirLocationDao.searchForLocations(null, null, null, null, null, locationTag, null,
+		    null);
+		
 		assertThat(locations, notNullValue());
-		assertThat(locations.size(), greaterThanOrEqualTo(2));
+		assertThat(locations.size(), equalTo(2));
```
Comment (VaishSiddharth) [inline api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Here something like this might work

```java 
		assertThat(locations.iterator().next().getTags().iterator().next().getLocationTagId(),equalTo(LOCATIONT_TAG));
```

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 15]
```diff
@@ -101,91 +113,140 @@ public void getLocationByUuid_shouldReturnNullWithUnknownUuid() {
...
-	public void findLocationByCity_shouldReturnCorrectLocation() {
-		Collection<Location> locations = fhirLocationDao.findLocationsByCity(LOCATION_CITY);
+	public void searchForLocations_shouldReturnCorrectLocationByCity() {
+		StringParam city = new StringParam();
+		city.setValue(LOCATION_CITY);
+		Collection<Location> locations = fhirLocationDao.searchForLocations(null, city, null, null, null, null, null, null);
 		
 		assertThat(locations, notNullValue());
-		assertThat(locations.size(), greaterThanOrEqualTo(1));
+		assertThat(locations.size(), equalTo(1));
```
[the remark this answers is not in the document]
Comment (CaptainDaVinci) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: I've noticed these in few other tests as well, shouldn't every resource item be tested to see if contains the correct value according to the search query?

File: omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java [hunk 16]
```diff
@@ -56,37 +58,13 @@ public Location getLocationById(@IdParam @NotNull IdType id) {
...
-	
-	@Search
-	@SuppressWarnings("unused")
-	public Bundle findLocationsByTag(@RequiredParam(name = "_tag") TokenParam tag) {
-		return FhirUtils.convertSearchResultsToBundle(fhirLocationService.findLocationsByTag(tag));
+	public Bundle searchLocations(@OptionalParam(name = Location.SP_NAME) StringParam name,
+	        @OptionalParam(name = Location.SP_ADDRESS_CITY) StringParam city,
+	        @OptionalParam(name = Location.SP_ADDRESS_COUNTRY) StringParam country,
+	        @OptionalParam(name = Location.SP_ADDRESS_POSTALCODE) StringParam postalCode,
+	        @OptionalParam(name = Location.SP_ADDRESS_STATE) StringParam state, @OptionalParam(name = "_tag") TokenParam tag,
```
Comment (CaptainDaVinci) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java]: You can replace the literal `"_tag"` by `PARAM_TAG` present in `ca.uhn.fhir.rest.api.Constants`

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (VaishSiddharth) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Yes every test should check according to query where every possible

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java (hunk 9 above)
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: I was stuck on this for a while because we need to use "Subqueries.propertyIn" and not "Subqueries.in" as the test was failing repeatedly. Thanks for the hint though :)

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java [hunk 17]
```diff
@@ -470,6 +474,28 @@ protected void handleLocationReference(Criteria criteria, ReferenceParam locatio
...
+					case Location.SP_ADDRESS_COUNTRY:
+						return propertyLike("country", parentReference.getValue());
+					case Location.SP_ADDRESS_POSTALCODE:
+						return propertyLike("postalCode", parentReference.getValue());
+					case "":
+						return Optional.of(eq("locationId", Integer.parseInt(parentReference.getValue())));
+				}
+			}
+		}
+		return Optional.empty();
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: I prefer to have a space before the `return` line.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 18]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
 	
 	@Override
-	public Collection<Location> findLocationByName(String name) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("name", name)).list();
+	public Collection<Location> searchForLocations(StringOrListParam name, StringOrListParam city, StringOrListParam country,
+	        StringOrListParam postalCode, StringOrListParam state, TokenOrListParam tag, ReferenceOrListParam parent,
+	        SortSpec sort) {
+		
+		Criteria criteria = this.sessionFactory.getCurrentSession().createCriteria(Location.class);
+		handleBooleanProperty("retired", false).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: I'd group these with the other `handle...` calls.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 19]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+		        .createAlias("attributeType", "lat").add(eq("lat.uuid", locationAttributeTypeUuid)).add(eq("voided", false))
+		        .list();
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByCountry(String country) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("country", country)).list();
+	private void handleName(Criteria criteria, StringOrListParam namePattern) {
+		if (namePattern != null) {
+			handleOrListParam(namePattern, (givenName) -> propertyLike("name", givenName)).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: `givenName` really only applies to things with a `PersonName` (OpenMRS) or `HumanName` (FHIR). This should just be called `name`, i.e., it gets called "givenName" to distinguish it from "familyName" (it makes a bit more sense in French where they're, respectively, "prénom" and "nom"). Very European-parochial.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 20]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+		}
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByPostalCode(String postalCode) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("postalCode", postalCode)).list();
+	private void handleCity(Criteria criteria, StringOrListParam city) {
+		if (city != null) {
+			handleOrListParam(city, (givenCity) -> propertyLike("cityVillage", givenCity)).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: Ditto on removing `given`

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 21]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+		}
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByState(String state) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("stateProvince", state)).list();
+	private void handleCountry(Criteria criteria, StringOrListParam country) {
+		if (country != null) {
+			handleOrListParam(country, (givenCountry) -> propertyLike("country", givenCountry)).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: Ditto given

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 22]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+	private void handlePostalCode(Criteria criteria, StringOrListParam postalCode) {
+		if (postalCode != null) {
+			handleOrListParam(postalCode, (givenPostalCode) -> propertyLike("postalCode", givenPostalCode))
+			        .ifPresent(criteria::add);
+		}
+	}
+	
+	private void handleState(Criteria criteria, StringOrListParam state) {
+		if (state != null) {
+			handleOrListParam(state, (givenState) -> propertyLike("stateProvince", givenState)).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: Ditto given

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 23]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+		}
+	}
+	
+	private void handleTag(Criteria criteria, TokenOrListParam tag) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: The `tag` parameter should be called `tags`

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java [hunk 24]
```diff
@@ -42,43 +55,97 @@ public Location getLocationByUuid(String uuid) {
...
+		if (tag != null) {
+			criteria.createAlias("tags", "t");
+			handleOrListParam(tag, (givenTag) -> Optional.of(eq("t.name", givenTag.getValue()))).ifPresent(criteria::add);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java]: Ditto given.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java [hunk 25]
```diff
@@ -470,6 +474,28 @@ protected void handleLocationReference(Criteria criteria, ReferenceParam locatio
 		}
 	}
 	
+	protected Optional<Criterion> handleParentReference(ReferenceParam parentReference) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: Why is this here instead of in the `LocationDaoImpl`? Unlike other reference handlers, this is pretty specific to `Location`s.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java [hunk 26]
```diff
@@ -470,6 +474,28 @@ protected void handleLocationReference(Criteria criteria, ReferenceParam locatio
...
+					case Location.SP_ADDRESS_CITY:
+						return propertyLike("cityVillage", parentReference.getValue());
+					case Location.SP_ADDRESS_STATE:
+						return propertyLike("stateProvince", parentReference.getValue());
+					case Location.SP_ADDRESS_COUNTRY:
+						return propertyLike("country", parentReference.getValue());
+					case Location.SP_ADDRESS_POSTALCODE:
+						return propertyLike("postalCode", parentReference.getValue());
+					case "":
+						return Optional.of(eq("locationId", Integer.parseInt(parentReference.getValue())));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: We shouldn't expect a `locationId` passed in here. We should expect a `uuid` for the location instead. Generally speaking, except for the special case of concepts, we should only be dealing with UUIDs (concepts are special for historical reasons).

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: I concur with @VaishSiddharth here and elsewhere. We should be checking that the property we're querying by is in the actual results.

File: omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java [hunk 27]
```diff
@@ -9,18 +9,23 @@
  */
 package org.openmrs.module.fhir2.providers;
 
+import static ca.uhn.fhir.rest.api.Constants.PARAM_TAG;
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProvider.java]: I wouldn't import this statically for consistency with the other parameters.

File: omod/src/test/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProviderTest.java [hunk 28]
```diff
@@ -129,124 +121,22 @@ public void getLocationWithWrongUuid_shouldThrowResourceNotFoundException() {
 	}
 	
 	@Test
-	public void findLocationsByName_shouldReturnMatchingBundleOfLocations() {
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProviderTest.java]: We actually do want most of these tests, even if they need to be modified.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirTaskDaoImpl.java [hunk 29]
```diff
@@ -78,7 +78,8 @@ public FhirTask getTaskByUuid(String uuid) {
 		// TODO: Handle optional params
 		// Task.basedOn
 		if (validReferenceParam(basedOnReference)) {
-			criteria.createAlias("basedOnReferences", "bo").add(Restrictions.eq("bo.reference", basedOnReference.getIdPart()))
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirTaskDaoImpl.java]: Don't include this in this PR

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 30]
```diff
@@ -95,91 +118,201 @@ public void getLocationByUuid_shouldReturnNullWithUnknownUuid() {
...
 		assertThat(locations, notNullValue());
-		assertThat(locations.size(), greaterThanOrEqualTo(2));
+		assertThat(locations.size(), equalTo(1));
+		assertThat(locations.iterator().next().getParentLocation().getUuid(), equalTo(LOCATION_PARENT_ID));
+	}
+	
+	@Test
+	public void searchForLocations_shouldReturnCorrectLocationByParentName() {
+		ReferenceOrListParam parentLocation = new ReferenceOrListParam()
+		        .add(new ReferenceParam().setChain("name").setValue(LOCATION_PARENT_NAME));
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: This should actually be `.setValue(...).setChain(...)` and so on for the others. The reason is that the `.setValue()` method resets the value of the chain. This will probably fix the coverage.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirTaskDaoImpl.java (hunk 29 above)
Comment (varung-31) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirTaskDaoImpl.java]: I really don't know how it came here!

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java [hunk 31]
```diff
@@ -192,4 +325,140 @@ public void getActiveAttributesByLocationAndAttributeTypeUuid_shouldReturnLocati
...
+			case "name":
+				locationList.removeIf(p -> p.getName() == null);
+				break;
+			case "address-city":
+				locationList.removeIf(p -> p.getCityVillage() == null);
+				break;
+			case "address-state":
+				locationList.removeIf(p -> p.getStateProvince() == null);
+				break;
+			case "address-postalCode":
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: `address-postalcode`

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java (hunk 30 above)
Comment (varung-31) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: This did fix the coverage! Thanks.

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java (hunk 31 above)
Comment (varung-31) [reply api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImplTest.java]: Yeah I noticed that because of the coverage :P
